Analytics routes

simzzz opened the pull request:
To-Do:

- [x] Discuss if we are going to just leave query start/end fields to be milliseconds

simzzz committed:
initial progress

simzzz committed:
Merge branch 'aip-61-adex-v5' into analytics-routes

simzzz committed:
initial implementation for analytics v5 routes finished

simzzz committed:
code clean up

elpiel commented on docs/config/ganache.toml:
```diff
@@ -41,7 +41,7 @@ ethereum_adapter_relayer = 'http://localhost:8888'
 creators_whitelist = []
 validators_whitelist = []
-
+admins = ['0xce07CbB7e054514D590a0262C93070D838bFBA2e']
```
You can use the real ganache ID.
Find it here:
https://github.com/AmbireTech/adex-validator-stack-rust/blob/bfdc4f75df7783350bc67ce5418ad0a79c747c93/primitives/src/test_util.rs?_pjax=%23js-repo-pjax-container%3Afirst-of-type%2C%20div%5Bitemtype%3D%22http%3A%2F%2Fschema.org%2FSoftwareSourceCode%22%5D%20main%3Afirst-of-type%2C%20%5Bdata-pjax-container%5D%3Afirst-of-type#L8

elpiel commented on primitives/src/analytics.rs:
```diff
@@ -91,6 +154,34 @@ pub enum OperatingSystem {
+}
+
+impl Timeframe {
+    pub fn get_period_in_hours(&self) -> i64 {
```
keep it simple: `to_hours`, also consider using u32, u64

simzzz replied:
it's always positive and it's used in `Duration::hours()` which takes i64

elpiel commented on primitives/src/sentry.rs:
```diff
@@ -203,6 +203,14 @@ pub struct Analytics {
+pub struct FetchedAnalytics {
+    pub payout_amount: Option<UnifiedNum>,
+    pub payout_count: Option<u32>,
+}
```
```
.then(aggr => ({
			limit: appliedLimit,
			limitReached: appliedLimit === aggr.length,
			aggr
		}))
```

While the aggr is:

```
const resultProjection = {
		value: '$value',
		time: {
			$toLong: {
				$dateFromParts: {
					year: { $ifNull: ['$_id.time.year', 0] },
					month: { $ifNull: ['$_id.time.month', 1] },
					day: { $ifNull: ['$_id.time.day', 1] },
					hour: { $ifNull: ['$_id.time.hour', 0] },
					timezone: timezone || 'UTC'
				}
			}
		},
		segment: '$_id.segment',
		_id: 0
	}
```
You can read what each part does, like `toLong` in Mongo docs to see what we should return at the end:
https://docs.mongodb.com/manual/reference/operator/aggregation/toLong/

It will be even better if you double check what the JS implementation returns as a json at the end.

elpiel commented on sentry/src/db/analytics.rs:
```diff
@@ -1,230 +1,73 @@
+    end_date: Option<DateHour<Utc>>,
+    query: &AnalyticsQuery,
+    auth_as_key: Option<String>,
+    auth_uid: ValidatorId,
```
You can use an enum for this. `AuthenticateAs::Advertiser(Address)`, ``AuthenticateAs::Publisher(Address)``

elpiel commented on sentry/src/db/analytics.rs:
```diff
@@ -1,230 +1,73 @@
+        None => (
+            format!("{}, time", query.metric.to_string()),
+            "time".to_string(),
+        ),
```
Use a vec for each of the clauses. This complicated query needs more than just strings.
You can still build the 2 vec based on the segment if you want to keep the ordering of the clauses (for groups in particular):
```
let mut select = vec!["time".to_string(), query.metric.to_string()];
let mut order_by = vec!["time".to_string()];

match Some(segment_by) = segment_by {
select.push(...);
order_by.push(...);
}
```

When building select SQL:

select.join(',')
order_bu.join(',')

elpiel commented on sentry/src/db/analytics.rs:
```diff
@@ -1,230 +1,73 @@
     where_clauses.extend(vec![
-        format!("event_type = ${}", params.len()),
-        format!("{} IS NOT NULL", metric),
+        format!("event_type = ${}", query.event_type),
```
$1 $2 $3 is for prepared statements.
You've removed `params.len()` so the number for the parameter in the SQL query will be wrong.

elpiel commented on primitives/src/analytics.rs:
```diff
@@ -76,10 +77,72 @@ pub struct AnalyticsQuery {
-    pub segment_by_channel: Option<String>,
+    pub timeframe: Timeframe,
+    pub segment_by: Option<String>,
+    pub start: Option<DateTime<Utc>>,
+    pub end: Option<DateTime<Utc>>,
```
We should probably deserialize directly into DateHour.

Also, the `Joi` package usually supports both date (String) & milliseconds (number) but we have to double check if we are going to just leave it to milliseconds (probably with @Ivshti ).

We use milliseconds almost everywhere in the JSON representaiton.

for more details check:
https://joi.dev/api/?v=17.5.0#date

elpiel commented on primitives/src/analytics.rs:
```diff
@@ -76,10 +77,72 @@ pub struct AnalyticsQuery {
+    pub end: Option<DateTime<Utc>>,
+    // #[serde(default = "default_timezone")]
+    // pub timezone: String,
+    pub campaign_id: Option<String>,
```
this is not a string.

elpiel commented on primitives/src/analytics.rs:
```diff
@@ -76,10 +77,72 @@ pub struct AnalyticsQuery {
+    // #[serde(default = "default_timezone")]
+    // pub timezone: String,
+    pub campaign_id: Option<String>,
+    pub ad_unit: Option<String>,
+    pub ad_slot: Option<String>,
```
`IPFS`

elpiel commented on primitives/src/analytics.rs:
```diff
@@ -76,10 +77,72 @@ pub struct AnalyticsQuery {
+    pub ad_slot: Option<String>,
+    pub ad_slot_type: Option<String>,
+    pub advertiser: Option<String>,
+    pub publisher: Option<String>,
```
`Address`

elpiel commented on primitives/src/analytics.rs:
```diff
@@ -76,10 +77,72 @@ pub struct AnalyticsQuery {
+    pub publisher: Option<String>,
+    pub hostname: Option<String>,
+    pub country: Option<String>,
+    pub os_name: Option<String>,
```
`OperatingSystem`

simzzz committed:
(still WIP) requested changes + test coverage

simzzz committed:
modified output + output is now split by timeframe

elpiel committed:
primitives - analytics - rename_all=camelCase

simzzz committed:
improved query, small changes to logic

simzzz committed:
Merge branch 'analytics-routes' of https://github.com/AmbireTech/adex-validator-stack-rust into analytics-routes

elpiel committed:
improvements on analytics

elpiel committed:
fix impl Display for DateHour

simzzz committed:
analytics query can now accept both timestamps and date strings

simzzz committed:
tests almost ready + changes to logic

simzzz committed:
Additional tests

simzzz committed:
test progress

simzzz committed:
fixed failing tests and logic + added myself as author

simzzz committed:
output is represented correctly + other minor changes

elpiel committed:
Cargo - patch postgres-types

elpiel committed:
primitives - senty:
- FetchedAnalytics - use DateTime
- test impl DateHour - &Timeframe

elpiel committed:
Cargo patch postgres-types:
Use new branch `from-T-for-boxed-dyn-ToSql`

elpiel committed:
primitives - analytics - enum AllowedKeys & struct Time

elpiel commented:
TODOs:
- [ ] Check (aka test) impls for Ord, PartialOrd, PartialEq for DateHour

elpiel committed:
adapter - fix dummy warning

elpiel committed:
Analytics changes:
- primitives - DateHour - PartialEq/PartialOrd
- primitives - AllowedKey impls & use Time
- sentry - migrations - rename column `os` to `os_name`
- sentry - db - analytics - improve SQL query and prepare testing

elpiel committed:
sentry - Event improvements & new enum EventType

elpiel committed:
sentry - db - analytics testing:
- Use primitives::sentry::EventType in structs
- Use primitives::analytics::FetchedMetric
- primitives::UnifiedNum - add Mul impls for u64
- sentry - db - analytics - finish query & testing

elpiel committed:
sentry - routes - analytics - finish route & tests

elpiel committed:
Merge branch 'aip-61-adex-v5' into analytics-routes

simzzz commented:
LGTM, I think the changes reduce the complexity of the logic we had in some places.

elpiel commented:
> TODOs:
> 
>     * [x]  Check (aka test) impls for Ord, PartialOrd, PartialEq for DateHour
> 
>     * [x]  More tests for `query::Time`
> 
>     * [ ]  Deserialize DateHour from either timestamp or ISO format
> 
>     * [ ]  Caching analytics (if needed)

I'm moving the last 2 TODOs to the #382 issue, these can be implemented on later stage.
